### PR TITLE
Allow interfaces to be declared and implemented within intrinsics

### PR DESCRIPTION
**Pull request by gusty**
This was discussed as part of https://github.com/fsharp/fslang-suggestions/issues/691

### CODE REVIEW

**Comment by gusty** [comment]
@KevinRansom I don't understand why I'm still getting this error stating that there's a difference in neg46 when I already removed the relevant part from the neg46.bsl

**Comment by cartermp** [comment]
@gusty if you go here (for some reason not working on my mac laptop) and click on fsharpQA logs you can see detailed failures: https://dev.azure.com/dnceng/public/_build/results?buildId=1030986&view=artifacts&pathAsName=false&type=publishedArtifacts

**Comment by gusty** [comment]
So far here we're removing the limitation.

Another PR could adjust the warnings as suggested by @dsyme and remove it completely when the containing module is a rec one.

Please consider including this in the next preview as we need this in our projects to separate the business logic (at least in the same file at the top).

**Comment by Happypig375** [comment]
Bump @dsyme

**Comment by dsyme** [comment]
@gusty The PR doesn't match what's approved?

> The limitation ...[is to be]....turned into a warning when "namespace rec" is not used, and removed altogether when it is.

**Comment by gusty** [comment]
@dsyme as I said, it doesn't remove the warning in the `rec` case, so yes it doesn't match 100% but it fails (warns) on the safe side.

Anyway, I can have another look at it and try to suppress the warning in those cases, last time I checked I was unable to figure out where is that module information available.

**Comment by dsyme** [comment]
@gusty The PR doesn't add any tests showing what is now allowed

To be clearer, the approved suggestion is

1. The warning message

       a.fs(8,14): warning FS0069: Interface implementations in augmentations are now deprecated. Interface implementations should be given on the initial declaration of a type

   gets changed to

       a.fs(8,14): warning FS0069:  Interface implementations should normally be given on the initial declaration of a type. Interface implementations in augmentations may lead to accessing static bindings before they are initialized, though only if the interface implementation is invoked during initialization of the static data, and in turn access the static data.  You may remove this warning using #nowarn "69" if you have checked this is not the case."


2. The warning message is suppressed if `namesapce rec` is used

It's ok to just do part (1), since it;s only a warning message.

**Comment by T-Gro** [comment]
@gusty Hi, fine if I take this over ?

**Comment by gusty** [comment]
@T-Gro Sure, go ahead.

It's basically finished, just need to remove the warning in the mentioned case.
I didn't remove it as it requires some information about the module which is not directly available so have to figure out where can I pull it from.

**Comment by T-Gro** [comment]
@gusty  ; @dsyme  ; @vzarytovskii 

Checks passing now, ready to be reviewed.

File: src/Compiler/Checking/CheckDeclarations.fs [hunk 1]
```diff
@@ -4700,7 +4701,7 @@ let rec TcModuleOrNamespaceElementNonMutRec (cenv: cenv) parent typeNames scopem
           // Treat 'module rec M = ...' as a single mutually recursive definition group 'module M = ...'
           if isRec then 
               assert (not isContinuingModule)
-              let modDecl = SynModuleDecl.NestedModule(compInfo, false, moduleDefs, isContinuingModule, m, trivia)
+              let modDecl = SynModuleDecl.NestedModule(compInfo, false, moduleDefs, isContinuingModule, m, trivia)
```
Comment (gusty) [inline src/Compiler/Checking/CheckDeclarations.fs:4704]: nitpick: remove the additional spaces to minimize the diff

File: tests/FSharp.Compiler.ComponentTests/ErrorMessages/InterfaceImplInAugmentationsTests.fs [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+    | Data of string
+    interface System.IDisposable
+
+type MyCustomType<'T> with
+    interface System.IDisposable with
+                member x.Dispose() = printfn "Finish"   """
+        |> typecheck
+        |> shouldFail
+        |> withSingleDiagnostic (Warning 69, Line 8, Col 15, Line 8, Col 33,
+                                 """Interface implementations should normally be given on the initial declaration of a type. Interface implementations in augmentations may lead to accessing static bindings before they are initialized, though only if the interface implementation is invoked during initialization of the static data, and in turn access the static data.  You may remove this warning using #nowarn "69" if you have checked this is not the case.""")
```
Comment (gusty) [inline tests/FSharp.Compiler.ComponentTests/ErrorMessages/InterfaceImplInAugmentationsTests.fs:24]: Isn't it worth storing this in a const literal at the beginning of the file? so if the message is eventually adjusted there's no massive replaces across tests.

**Review by gusty — COMMENTED** [review]
Looks good to me, I can't approve as I'm considered as author. Thanks for taking care of this.

File: tests/FSharp.Compiler.ComponentTests/ErrorMessages/InterfaceImplInAugmentationsTests.fs (hunk 2 above)
Comment (smoothdeveloper) [reply tests/FSharp.Compiler.ComponentTests/ErrorMessages/InterfaceImplInAugmentationsTests.fs:24]: Always felt those tests are better done in standalone F# files with baseline files of compiler output.

* It makes it easier to update the error message, compare the changes without "run tests / edit / recompile / ..." cycles, just by updating the baselines and comparing the diff through git tooling.
* looking at the F# code being checked makes more sense in standalone F# file than inside a string literal

Comment (T-Gro) [reply tests/FSharp.Compiler.ComponentTests/ErrorMessages/InterfaceImplInAugmentationsTests.fs:24]: > Isn't it worth storing this in a const literal at the beginning of the file? so if the message is eventually adjusted there's no massive replaces across tests.

In the end changes of error messages are bulk find+replace anyway, since there are various suites of tests that use different mechanisms.

That way, it does not matter for replacing.
And for searching, it is at least one level of indirection less.

But I am rather agnostic, I also thought about referencing the text from the resource (where the compiler takes it from)....

Comment (T-Gro) [reply tests/FSharp.Compiler.ComponentTests/ErrorMessages/InterfaceImplInAugmentationsTests.fs:24]: > Always felt those tests are better done in standalone F# files with baseline files of compiler output.
> 
> * It makes it easier to update the error message, compare the changes without "run tests / edit / recompile / ..." cycles, just by updating the baselines and comparing the diff through git tooling.
> * looking at the F# code being checked makes more sense in standalone F# file than inside a string literal

As soon as I had to look at several tests at once (trying different approaches and having different subsets failing/passing), having all the tests and assertions was a massive benefit.
For a feature having a single test, separate .fs is better if the code is non-trivial (=difficult to understand in a string-only setting).
For a suite of several smaller tests, the ability to see them (and their assertions) at glance is the selling point.

File: src/Compiler/FSStrings.resx [hunk 3]
```diff
@@ -1054,7 +1054,7 @@
     <value>Override implementations should be given as part of the initial declaration of a type.</value>
   </data>
   <data name="IntfImplInIntrinsicAugmentation" xml:space="preserve">
-    <value>Interface implementations in augmentations are now deprecated. Interface implementations should be given on the initial declaration of a type.</value>
+    <value>Interface implementations should normally be given on the initial declaration of a type. Interface implementations in augmentations may lead to accessing static bindings before they are initialized, though only if the interface implementation is invoked during initialization of the static data, and in turn access the static data.  You may remove this warning using #nowarn "69" if you have checked this is not the case.</value>
```
Comment (dsyme) [inline src/Compiler/FSStrings.resx:1057]: Minor nit - two spaces after `.` - my understanding is we only use one.

File: src/Compiler/Checking/CheckDeclarations.fs [hunk 4]
```diff
@@ -4139,7 +4140,7 @@ module TcDeclarations =
     //-------------------------------------------------------------------------
 
     /// Bind a collection of mutually recursive definitions in an implementation file
-    let TcMutRecDefinitions (cenv: cenv) envInitial parent typeNames tpenv m scopem mutRecNSInfo (mutRecDefns: MutRecDefnsInitialData) =
+    let TcMutRecDefinitions (cenv: cenv) envInitial parent typeNames tpenv m scopem mutRecNSInfo (mutRecDefns: MutRecDefnsInitialData) skipExternalAugmWarning =
```
Comment (dsyme) [inline src/Compiler/Checking/CheckDeclarations.fs:4143]: I'd rather `skipExternalAugmWarning` was called `isMutRec`.  I believe that's accurate and the flag may have other similar uses later.

File: src/Compiler/Checking/CheckDeclarations.fs [hunk 5]
```diff
@@ -1650,11 +1650,12 @@ let TcMutRecDefns_Phase2 (cenv: cenv) envInitial mBinds scopem mutRecNSInfo (env
...
                       (generatedHashAndEqualsWithComparerValues && typeEquiv g intfTyR g.mk_IStructuralEquatable_ty) then
                       errorR(Error(FSComp.SR.tcDefaultImplementationForInterfaceHasAlreadyBeenAdded(), intfTy.Range))
-
-                  if overridesOK = WarnOnOverrides then  
-                      warning(IntfImplInIntrinsicAugmentation(intfTy.Range))
-                  if overridesOK = ErrorOnOverrides then  
-                      errorR(IntfImplInExtrinsicAugmentation(intfTy.Range))
+           
+                  match skipExternalAugmWarning, overridesOK with
+                  | _, OverridesOK | true, WarnOnOverrides -> ()
```
Comment (dsyme) [inline src/Compiler/Checking/CheckDeclarations.fs:1655]: Split out these two cases, and please add a comment on each clause of this match describing what's gonig on in plain English.

**Review by dsyme — COMMENTED** [review]
Just a couple of minor nits.  Looks great!

I'll mark as "comment" to not block approval - consider it approved besides these minor nits

File: tests/FSharp.Compiler.ComponentTests/ErrorMessages/InterfaceImplInAugmentationsTests.fs [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+
+namespace FSharp.Compiler.ComponentTests.ErrorMessages
+
+open Xunit
+open FSharp.Test.Compiler
+
+
+module InterfaceImplAugmentation =
+    [<Fact>]
+    let ``Type in non-recursive namespace show give a warning when augmented externally``() =
```
Comment (0101) [inline tests/FSharp.Compiler.ComponentTests/ErrorMessages/InterfaceImplInAugmentationsTests.fs:11]:
```suggestion
    let ``Type in non-recursive namespace gives a warning when augmented externally``() =
```

File: tests/FSharp.Compiler.ComponentTests/ErrorMessages/InterfaceImplInAugmentationsTests.fs [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+type MyCustomType<'T> with
+    interface System.IDisposable with
+                member x.Dispose() = printfn "Finish"   """
+        |> typecheck
+        |> shouldFail
+        |> withSingleDiagnostic (Warning 69, Line 8, Col 15, Line 8, Col 33,
+                                 """Interface implementations should normally be given on the initial declaration of a type. Interface implementations in augmentations may lead to accessing static bindings before they are initialized, though only if the interface implementation is invoked during initialization of the static data, and in turn access the static data. You may remove this warning using #nowarn "69" if you have checked this is not the case.""")
+
+    [<Fact>]
+    let ``Exception type in non-recursive namespace show give a warning when augmented externally``() =
```
Comment (0101) [inline tests/FSharp.Compiler.ComponentTests/ErrorMessages/InterfaceImplInAugmentationsTests.fs:27]:
```suggestion
    let ``Exception type in non-recursive namespace gives a warning when augmented externally``() =
```

**Review by 0101 — DISMISSED** [review]

**Review by 0101 — APPROVED** [review]

**Comment by dsyme** [comment]
@T-Gro I updated the branch here

**Review by dsyme — APPROVED** [review]

**Comment by dsyme** [comment]
@vzarytovskii This could get mentioned in F# 7 release announcement, perhaps also a very brief RFC that the limitation has been lifted.
